src/kv/untyped_map.h

```diff
@@ -0,0 +1,734 @@
+          // If we depend on the key not existing, it must be absent.
+          if (search.has_value())
+          {
+            LOG_DEBUG_FMT("Read depends on non-existing entry");
+            return false;
+          }
+        }
+        else
+        {
+          // If we depend on the key existing, it must be present and have the
+          // version that we expect.
+          if (!search.has_value() || (it->second != search.value().version))
+          {
+            LOG_DEBUG_FMT("Read depends on invalid version of entry");
+            return false;
+          }
+        }
+      }
+
+      return true;
+    }
+
+    void commit(Version v) override
+    {
+      if (change_set.writes.empty())
+      {
+        commit_version = change_set.start_version;
+        return;
+      }
+
+      // Record our commit time.
+      commit_version = v;
+      committed_writes = true;
+
+      if (!change_set.writes.empty())
+      {
+        auto& roll = map.get_roll();
+        auto state = roll.commits->get_tail()->state;
+
+        for (auto it = change_set.writes.begin(); it != change_set.writes.end();
+             ++it)
+        {
+          if (it->second.has_value())
+          {
+            // Write the new value with the global version.
+            changes = true;
+            state = state.put(it->first, VersionV{v, it->second.value()});
+          }
+          else
+          {
+            // Write an empty value with the deleted global version only if
+            // the key exists.
+            auto search = state.get(it->first);
+            if (search.has_value())
+            {
+              changes = true;
+              state = state.put(it->first, VersionV{-v, {}});
+            }
+          }
+        }
+
+        if (changes)
+        {
+          map.append_to_roll(v, state, change_set.writes);
+        }
+      }
+    }
+
+    void post_commit() override
+    {
+      // This is run separately from commit so that all commits in the Tx
+      // have been applied before local hooks are run. The maps in the Tx
+      // are still locked when post_commit is run.
+      if (change_set.writes.empty())
+        return;
+
+      map.trigger_local_hook();
+    }
+
+    // Used by owning map during serialise and deserialise
+    ChangeSet& get_change_set()
+    {
+      return change_set;
+    }
+
+    const ChangeSet& get_change_set() const
+    {
+      return change_set;
+    }
+
+    void set_commit_version(Version v)
+    {
+      commit_version = v;
+    }
+  };
+
+  struct ConcreteTxView : public TxViewCommitter, public TxView
+  {
+    ConcreteTxView(
+      RollContainer& m,
+      size_t rollbacks,
+      State& current_state,
+      State& committed_state,
+      Version v) :
+      TxViewCommitter(m, rollbacks, current_state, committed_state, v),
+      TxView(TxViewCommitter::change_set)
+    {}
+  };
+
+  class Map : public AbstractMap, public RollContainer
+  {
+  public:
+    using K = SerialisedEntry;
```

Comment:
Although the Key and Value are the same type, I've retained the K and V names so its clear which we _intend_ to use within this class.